opal/mca/btl/ofi/btl_ofi_component.c

```diff
@@ -0,0 +1,583 @@
+    struct fid_cq *cq;
+    struct fid_av *av;
+
+    /* make a copy of the given info to store on the module */
+    ofi_info = fi_dupinfo(info);
+
+    linux_device_name = info->domain_attr->name;
+    BTL_VERBOSE(("initializing dev:%s provider:%s",
+                    linux_device_name,
+                    info->fabric_attr->prov_name));
+
+    /* fabric */
+    rc = fi_fabric(ofi_info->fabric_attr, &fabric, NULL);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_fabric with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+    /* domain */
+    rc = fi_domain(fabric, ofi_info, &domain, NULL);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_domain with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+    /* endpoint */
+    rc = fi_endpoint(domain, ofi_info, &endpoint, NULL);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_endpoint with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+    /* CQ */
+    cq_attr.format = FI_CQ_FORMAT_CONTEXT;
+    cq_attr.wait_obj = FI_WAIT_NONE;
+    rc = fi_cq_open(domain, &cq_attr, &cq, NULL);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_cq_open with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+    /* AV */
+    av_attr.type = FI_AV_MAP;
+    rc = fi_av_open(domain, &av_attr, &av, NULL);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_av_open with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+
+    /* bind CQ and AV to endpoint */
+    uint32_t cq_flags = (FI_TRANSMIT);
+    rc = fi_ep_bind(endpoint, (fid_t)cq, cq_flags);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_ep_bind with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+    rc = fi_ep_bind(endpoint, (fid_t)av, 0);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_ep_bind with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+    /* enable the endpoint for using */
+    rc = fi_enable(endpoint);
+    if (0 != rc) {
+        BTL_VERBOSE(("%s failed fi_enable with err=%s",
+                        linux_device_name,
+                        fi_strerror(-rc)
+                        ));
+        goto fail;
+    }
+
+    /* Everything succeeded, lets create a module for this device. */
+    module = (mca_btl_ofi_module_t*) malloc(sizeof(mca_btl_ofi_module_t));
```

Comment:
I think you need to `calloc()` this (vs. `malloc()`) in order to ensure that the NULL protections against pointers in the module struct in the module finalize work as expected.